### ISSUE TITLE
how can i skip zlib build, and use system zlib library instead?

### ISSUE THREAD

**Issue by roygbip** [ClickHouse/ClickHouse#12709]
when build CH in my platform, meet this Error:

```shell
Building C object contrib/zlib-ng/CMakeFiles/zlib.dir/crc32.c.o
Assembler messages:
Error: unknown architectural extension `dotprod'
Error: unrecognized option -march=armv8.2-a+crypto+fp16+dotprod
make[2]: *** [contrib/zlib-ng/CMakeFiles/zlib.dir/compress.c.o] Error 2
make[2]: *** Waiting for unfinished jobs....
make[2]: *** [contrib/zlib-ng/CMakeFiles/zlib.dir/adler32.c.o] Error 2
make[2]: *** [contrib/zlib-ng/CMakeFiles/zlib.dir/crc32.c.o] Error 2
make[1]: *** [contrib/zlib-ng/CMakeFiles/zlib.dir/all] Error 2
```

arch: aarch64
dis: centos7
ch version:  v20.3.11.97

i download `yum install zlib-devel`
zlib:
```
/usr/include/zconf.h
/usr/include/zlib.h
/usr/lib64/libz.so
/usr/lib64/pkgconfig/zlib.pc
```

how can i skip zlib build, and use system zlib library instead

**Comment by roygbip** [comment]
change file `cmake/find/zlib.cmake` line 2:
```shell
if (NOT OS_FREEBSD AND NOT ARCH_32)
    option (USE_INTERNAL_ZLIB_LIBRARY "Set to FALSE to use system zlib library instead of bundled" ${NOT_UNBUNDLED})
endif ()
```
to
```shell
if (NOT OS_FREEBSD AND NOT ARCH_32)
    option (USE_INTERNAL_ZLIB_LIBRARY "Set to FALSE to use system zlib library instead of bundled" 0)
endif ()
```
worked

**Comment by roygbip** [comment]
```c++
ClickHouse/src/Functions/CRC.cpp:66:14: error: ‘crc32_z’ was not declared in this scope; did you mean ‘crc32’?
   66 |     { return crc32_z(0L, buf, size); }
```

**Comment by alexey-milovidov** [comment]
You can edit the file `ClickHouse/src/Functions/CRC.cpp` and disable CRC function if you build with system zlib.

**Comment by roygbip** [comment]
> You can edit the file `ClickHouse/src/Functions/CRC.cpp` and disable CRC function if you build with system zlib.

after remove CRC function, i meet:
```
during RTL pass: expand
In file included from /work/ClickHouse/src/Functions/bitRotateLeft.cpp:2:
/work/ClickHouse/src/Functions/FunctionBinaryArithmetic.h: In static member function ‘static void DB::BinaryOperationImplBase<A, B, Op, ResultType_>::vector_constant(const A*, B, DB::BinaryOperationImplBase<A, B, Op, ResultType_>::ResultType*, size_t) [with A = char8_t; B = long unsigned int; Op = DB::BitRotateLeftImpl<char8_t, long unsigned int>; ResultType_ = long unsigned int]’:
/work/ClickHouse/src/Functions/FunctionBinaryArithmetic.h:74:13: internal compiler error: in expand_shift_1, at expmed.c:2635
   74 |             c[i] = Op::template apply<ResultType>(a[i], b);
      |             ^
0x85819f expand_shift_1
	../../gcc/expmed.c:2635
0x8789af expand_expr_real_2(separate_ops*, rtx_def*, machine_mode, expand_modifier)
	../../gcc/expr.c:9368
0x8699fb expand_expr_real_1(tree_node*, rtx_def*, machine_mode, expand_modifier, rtx_def**, bool)
	../../gcc/expr.c:9943
0x871463 expand_expr
	../../gcc/expr.h:279
0x871463 expand_operands(tree_node*, tree_node*, rtx_def*, rtx_def**, rtx_def**, expand_modifier)
	../../gcc/expr.c:7874
0x876a6b expand_expr_real_2(separate_ops*, rtx_def*, machine_mode, expand_modifier)
	../../gcc/expr.c:9733
0x8699fb expand_expr_real_1(tree_node*, rtx_def*, machine_mode, expand_modifier, rtx_def**, bool)
	../../gcc/expr.c:9943
0x874073 expand_expr
	../../gcc/expr.h:279
0x874073 expand_assignment(tree_node*, tree_node*, bool)
	../../gcc/expr.c:5011
0x77aa9b expand_gimple_stmt_1
	../../gcc/cfgexpand.c:3752
0x77aa9b expand_gimple_stmt
	../../gcc/cfgexpand.c:3850
0x77fe0f expand_gimple_basic_block
	../../gcc/cfgexpand.c:5886
0x7821df execute
	../../gcc/cfgexpand.c:6509
Please submit a full bug report,
with preprocessed source if appropriate.
Please include the complete backtrace with any bug report.
See <https://gcc.gnu.org/bugs/> for instructions.
make[2]: *** [src/Functions/CMakeFiles/clickhouse_functions.dir/bitRotateLeft.cpp.o] Error 1
make[1]: *** [src/Functions/CMakeFiles/clickhouse_functions.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by roygbip** [comment]
change file `src/Functions/FunctionBinaryArithmetic.h`:73

```
 static void NO_INLINE vector_constant(const A * __restrict a, B b, ResultType * __restrict c, size_t size)
    {
        return;
        for (size_t i = 0; i < size; ++i)
            c[i] = Op::template apply<ResultType>(a[i], b);
    }
    
static void NO_INLINE constant_vector(A a, const B * __restrict b, ResultType * __restrict c, size_t size)
    {
        return;
        for (size_t i = 0; i < size; ++i)
            c[i] = Op::template apply<ResultType>(a, b[i]);
    }
```

works for me, but i don`t know if there are any impact.

**Comment by roygbip** [comment]
no kafka engine ?
```
Code: 56. DB::Exception: Received from localhost:9000. DB::Exception: Unknown table engine Kafka.
```

**Comment by roygbip** [comment]
> no kafka engine ?
> 
> ```
> Code: 56. DB::Exception: Received from localhost:9000. DB::Exception: Unknown table engine Kafka.
> ```

does kafka engine not support arrch64?

**Comment by alexey-milovidov** [comment]
> change file src/Functions/FunctionBinaryArithmetic.h:73

Don't do that, consequences will be dreadful.
(Arithmetic operations will return uninitialized memory, what can be worse.)

Internal compiler error means bug in compiler.
You are using gcc 9.1.0. It's old version as 9.3 is available and it contains many bugfixes.
This old version most likely comes from your package repository.
You should install fresh gcc instead.

Or you can switch to clang.
BTW, we are using clang to build aarch64 binaries.
We did not test whether gcc or clang produces better code on aarch64 (on amd64, gcc is slightly better).

**Comment by alexey-milovidov** [comment]
> does kafka engine not support arrch64?

Yes, Kafka is disabled on Aarch64, see `cmake/find/rdkafka.cmake`.
It's possible to make it work if we provide appropriate version of `contrib/librdkafka-cmake/config.h`.

**Comment by roygbip** [comment]
@alexey-milovidov  thx, that helps a lot.

**Comment by citrin** [comment]
> You can edit the file `ClickHouse/src/Functions/CRC.cpp` and disable CRC function if you build with system zlib.

Faced the same problem, what exactly should be changed in `CRC.cpp`?

**Comment by citrin** [comment]
I've just changed `crc32_z` to `crc32` - should fix the build with zlib < 1.2.9. It is highly unlikely that in our setup it would be used for strings >= 4Gb.

**Comment by filimonov** [comment]
> does kafka engine not support arrch64?

Support since: https://github.com/ClickHouse/ClickHouse/pull/19369
